File: Java/Language/Basics/src/Employee.java
package payroll;

public class Employee {

	private int id; //value will be separately initialized for each instance
	private int hours;
	private float rate;
	private static int count; //value will be initialized once and shared by all instances

	//parameterized constructor
	public Employee(int h, float r) {
		id = 101 + count++;
		hours = h;
		rate = r;
	}

	//parameterless constructor
	public Employee() {
		this(0, 50); // calling another constructor of this class
	}

	public int getId() { return id; }

	public int getHours() { return hours; }
	public void setHours(int value) { hours = value; }

	public float getRate() { return rate; }
	public void setRate(float value) { rate = value; }

	public double income() {
		double amount = hours * rate;
		int ot = hours - 180;
		if(ot > 0)
			amount += 50 * ot;
		return amount;
	}
}

File: CSharp/Basics/ClassTypeTest/DemoApp/Payroll/Employee.cs
namespace Payroll
{
	public class Employee
	{
		private int hours;
		private float rate;
		private int id;
		private static int count;

		public Employee()
		{
			id = 101 + count++;
		}

		public Employee(int hours, float rate) : this()
		{
			this.hours = hours;
			this.rate = rate;
		}

		public int Hours
		{
			get
			{
				return hours;
			}

			set
			{
				hours = value;
			}
		}

		public float Rate
		{
			get { return rate; }
			set { rate = value; }
		}

		public int Id
		{
			get { return id; }
		}

		public virtual double GetIncome()
		{
			double income = hours * rate;
			int ot = hours - 180;
			if(ot > 0)
				income += 50 * ot;
			return income;
		}

		public static int CountInstances()
		{
			return count;
		}
	}
}

